packages/firestore/test/unit/local/web_storage_metadata_channel.test.ts

```diff
@@ -0,0 +1,271 @@
+
+    it('with multiple batches', () => {
+      notificationChannel.addLocalPendingMutation(0);
+      notificationChannel.addLocalPendingMutation(1);
+      assertInstanceState([], 0, 1);
+
+      notificationChannel.addLocalPendingMutation(2);
+      notificationChannel.addLocalPendingMutation(3);
+      assertInstanceState([], 0, 3);
+
+      notificationChannel.removeLocalPendingMutation(0);
+      notificationChannel.removeLocalPendingMutation(2);
+      assertInstanceState([], 1, 3);
+    });
+  });
+
+  describe('persists query targets', () => {
+    beforeEach(() => {
+      return persistenceHelpers
+        .testLocalStorageNotificationChannel(ownerId, [], [])
+        .then(nc => {
+          notificationChannel = nc;
+        });
+    });
+
+    afterEach(() => {
+      notificationChannel.shutdown();
+    });
+
+    it('with empty targets', () => {
+      assertInstanceState([]);
+    });
+
+    it('with multiple targets', () => {
+      notificationChannel.addLocallyActiveQueryTarget(0);
+      assertInstanceState([0]);
+
+      notificationChannel.addLocallyActiveQueryTarget(1);
+      notificationChannel.addLocallyActiveQueryTarget(2);
+      assertInstanceState([0, 1, 2]);
+
+      notificationChannel.removeLocallyActiveQueryTarget(1);
+      assertInstanceState([0, 2]);
+    });
+  });
+
+  describe('combines instance data', () => {
+    let previousAddEventListener;
+    let storageCallback: (StorageEvent) => void;
+
+    beforeEach(() => {
+      previousAddEventListener = window.addEventListener;
+      window.addEventListener = (type, callback) => {
+        expect(type).to.be.equal('storage');
+        storageCallback = callback;
+      };
+
+      return persistenceHelpers
+        .testLocalStorageNotificationChannel(ownerId, [1, 2], [3, 4])
+        .then(nc => {
+          notificationChannel = nc;
+          expect(storageCallback).to.not.be.undefined;
+        });
+    });
+
+    afterEach(() => {
+      notificationChannel.shutdown();
+      window.addEventListener = previousAddEventListener;
+    });
+
+    const verifyState = (minBatchId: BatchId, expectedTargets: TargetId[]) => {
+      const actualTargets = notificationChannel.getGloballyActiveQueryTargets();
+      expect(actualTargets.size).to.equal(expectedTargets.length);
+
+      let missingTargets = actualTargets;
+      for (const targetId of expectedTargets) {
+        missingTargets = missingTargets.delete(targetId);
+      }
+      expect(missingTargets.size).to.equal(0);
+
+      expect(
+        notificationChannel.getMinimumGloballyPendingMutation()
+      ).to.be.equal(minBatchId);
+    };
+
+    it('with existing instance', () => {
+      // The prior instance has one pending mutation and two active query targets
+      verifyState(1, [3, 4]);
+
+      notificationChannel.addLocalPendingMutation(3);
+      notificationChannel.addLocallyActiveQueryTarget(4);
+      verifyState(1, [3, 4]);
+
+      notificationChannel.addLocalPendingMutation(0);
```

Comment:
this is technically invalid usage too... It should be >= 3

Reply:
Added comment